File: metac/vm.nim
import xrest, metac/fs, metac/rest_common, collections, metac/net, metac/desktop, options, metac/media

type
  DriveDriver* {.pure.} = enum
    virtio, ide

  Drive* = object
    driver*: DriveDriver
    device*: fs.FileRef

  BootKernel* = object
    kernel*: net.ByteStream
    initrd*: Option[net.ByteStream]
    cmdline*: string

  SerialPortDriver* {.pure.} = enum
    default, virtio

  AfterLaunch*[T] = Option[T]
    # used for parameters that are available only after launch

  SerialPort* = object
    driver*: SerialPortDriver
    name*: string
    nowait*: bool

  VMFilesystemDriver* = enum
    virtio9p

  VmFilesystem* = object
    driver*: VMFilesystemDriver
    name*: string
    fs*: FilesystemRef

  VmState* {.pure.} = enum
    running, turnedOff

  VM* = object
    meta*: Metadata
    state*: VmState
    memory*: int # in MiB
    vcpu*: int

    bootDisk*: Option[int]
    bootKernel*: Option[BootKernel]
    drives*: seq[Drive]
    filesystems*: seq[VmFilesystem]

    serialPorts*: seq[SerialPort]

restRef VMRef:
  get() -> VM
  sub("desktop", DesktopRef)
  update(VM)
  delete()
  # todo: support patch(VMPatch)

basicCollection(VM, VMRef)

File: py/metac/vm.py
from metac.core import *
from metac.fs import FilesystemRef, FileRef
from typing import NamedTuple, Optional
from enum import Enum

class DriveDriver(Enum):
    virtio = 1
    ide = 2

class Drive(NamedTuple):
    driver: DriveDriver
    device: FileRef

class BootKernel(NamedTuple):
    kernel: Ref
    cmdline: str
    initrd: Optional[Ref] = None

class SerialPortDriver(Enum):
    default = 1
    virtio = 2

class SerialPort(NamedTuple):
    driver: SerialPortDriver
    name: str
    nowait: bool

class VmFilesystemDriver(Enum):
    virtio9p = 1

class VmFilesystem(NamedTuple):
    name: str
    fs: FilesystemRef
    driver: VmFilesystemDriver = VmFilesystemDriver.virtio9p

class VmState(Enum):
    running = 1
    turnedOff = 2

class Vm(NamedTuple):
    meta: Metadata
    memory: int
    state: VmState = VmState.running
    vcpu: int = 1

    bootDisk: Optional[int] = None
    bootKernel: Optional[BootKernel] = None
    drives: List[Drive] = []
    filesystems: List[VmFilesystem] = []

    serialPorts: List[SerialPort] = []

class VmRef(Ref, GetMixin[Vm], DeleteMixin):
    value_type = Vm

class VmCollection(Ref, CollectionMixin[Vm, VmRef]):
    value_type = Vm
    ref_type = VmRef

def get_vms() -> VmCollection:
    return VmCollection('/vm/')

if __name__ == '__main__':
    vms = get_vms().values()
    get_vms().create(Vm(
        meta=Metadata(name='hello'),
        memory=1024,
    ))
    print(vms)
